diff --git a/src/EventSubscriber/RequireLoginSubscriber.php b/src/EventSubscriber/RequireLoginSubscriber.php
index 5fa8f82..26f31db 100644
--- a/src/EventSubscriber/RequireLoginSubscriber.php
+++ b/src/EventSubscriber/RequireLoginSubscriber.php
@@ -193,6 +193,11 @@ class RequireLoginSubscriber implements EventSubscriberInterface {
         $redirectPath = 'internal:/user/login';
       }
 
+      // Make sure base path is added to destination.
+      if ($base_path = $request->getBasePath()) {
+        $path = $base_path . $path;
+      }
+
       $redirect = Url::fromUri($redirectPath, ['query' => ['destination' => $path]])->toString();
       $event->setResponse(new ImmutableTargetRedirectResponse($redirect));
     }
